codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,333 @@
+                        State pState = state(parent);
+                        pState.offerAndInitializePseudoTime(state);
+                        pState.isActiveCountChangeForTree(state.activeCountForTree);
+                    }
+                }
+            }
+
+            @Override
+            public void onPriorityTreeParentChanging(Http2Stream stream, Http2Stream newParent) {
+                Http2Stream parent = stream.parent();
+                if (parent != null) {
+                    State state = state(stream);
+                    if (state.activeCountForTree != 0) {
+                        State pState = state(parent);
+                        pState.remove(state);
+                        pState.isActiveCountChangeForTree(-state.activeCountForTree);
+                    }
+                }
+            }
+        });
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState state) {
+        state(state.stream()).updateStreamableBytes(state.streamableBytes(),
+                                                    state.hasFrame() && state.isWriteAllowed());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        // As long as there is some active frame we should write at least 1 time.
+        if (connectionState.activeCountForTree == 0) {
+            return false;
+        }
+
+        // The goal is to write until we write all the allocated bytes or are no longer making progress.
+        // We still attempt to write even after the number of allocated bytes has been exhausted to allow empty frames
+        // to be sent. Making progress means the active streams rooted at the connection stream has changed.
+        int oldIsActiveCountForTree;
+        do {
+            oldIsActiveCountForTree = connectionState.activeCountForTree;
+            // connectionState will never be active, so go right to its children.
+            maxBytes -= distributeToChildren(maxBytes, writer, connectionState);
+        } while (connectionState.activeCountForTree != 0 &&
+                (maxBytes > 0 || oldIsActiveCountForTree != connectionState.activeCountForTree));
+
+        return connectionState.activeCountForTree != 0;
+    }
+
+    /**
+     * Sets the amount of bytes that will be allocated to each stream. Defaults to 1KiB.
+     * @param minAllocationChunk the amount of bytes that will be allocated to each stream. Must be > 0.
+     */
+    public void allocationQuantum(int allocationQuantum) {
+        if (allocationQuantum <= 0) {
+            throw new IllegalArgumentException("allocationQuantum must be > 0");
+        }
+        this.allocationQuantum = allocationQuantum;
+    }
+
+    private int distribute(int maxBytes, Writer writer, State state) throws Http2Exception {
+        if (state.active) {
+            if (maxBytes == 0 && state.streamableBytes > 0) {
```

Comment:
@nmittler - Actually I wanted to discuss this. We may be missing some information here. Streamable bytes could be 0 because the window is collapsed, and not just because there are no pending bytes to send (streamable bytes = `max(0, min(pendingBytes, window));`). So it is possible that we let a write slip because the window has collapsed and the stream has frames to send (example stream frame queue = `[Headers Frames(0 size), Data Frame(1Mb size)]`).

So can you clarify if my assumption for this check (don't write headers unless you can also write data) is what you had in mind. If this is the case I wonder if we should remove this check here and fix it as a followup PR for both of the `StreamByteDistributor` implementations...assuming we need to modify interfaces to accomplish this....WDYT?